File: JAVA/java/src/EvaluateBooleanBinaryTree.java
import java.util.LinkedList;
import java.util.Queue;

class Tree999 {
    int val;
    Tree999 left;
    Tree999 right;

    Tree999(int x) {
        val = x;
    }
}

public class EvaluateBooleanBinaryTree {

    public boolean evaluateTree(Tree999 root) {
        if (root.left == null && root.right == null) {
            // Handles the case for leaf nodes.
            return root.val != 0;
        }

        // Store the evaluations for the left subtree and right subtree.
        boolean evaluateLeftSubtree = evaluateTree(root.left);
        boolean evaluateRightSubtree = evaluateTree(root.right);

        if (root.val == 2) {
            return evaluateLeftSubtree || evaluateRightSubtree;
        } else {
            return evaluateLeftSubtree && evaluateRightSubtree;
        }
    }

    public Tree999 buildTreeFromList(Integer[] values) {
        if (values.length == 0) {
            return null;
        }

        Tree999 root = new Tree999(values[0]);
        Queue<Tree999> queue = new LinkedList<>();
        queue.add(root);
        int i = 1;

        while (i < values.length) {
            Tree999 current = queue.poll();
            if (values[i] != null) {
                current.left = new Tree999(values[i]);
                queue.add(current.left);
            }
            i++;
            if (i < values.length && values[i] != null) {
                current.right = new Tree999(values[i]);
                queue.add(current.right);
            }
            i++;
        }

        return root;
    }

    public static void main(String[] args) {
        EvaluateBooleanBinaryTree treeEvaluator = new EvaluateBooleanBinaryTree();

        // Input: [2, 1, 3, null, null, 0, 1]
        Integer[] inputArray = {2, 1, 3, null, null, 0, 1};

        Tree999 root = treeEvaluator.buildTreeFromList(inputArray);
        boolean result = treeEvaluator.evaluateTree(root);

        System.out.println(result);  // Output: true
    }
}

File: python/Evaluate Boolean Binary Tree.py
from typing import Optional, List

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def evaluateTree(root: Optional[TreeNode]) -> bool:
    if not root.left and not root.right:
        # Handles the case for leaf nodes.
        return root.val != 0

    # Store the evaluations for the left subtree and right subtree.
    evaluate_left_subtree = evaluateTree(root.left)
    evaluate_right_subtree = evaluateTree(root.right)
    if root.val == 2:
        evaluate_root = evaluate_left_subtree or evaluate_right_subtree
    else:
        evaluate_root = evaluate_left_subtree and evaluate_right_subtree

    return evaluate_root

def build_tree_from_list(values: List[Optional[int]]) -> Optional[TreeNode]:
    if not values:
        return None

    root = TreeNode(values[0])
    queue = [root]
    i = 1
    while i < len(values):
        current = queue.pop(0)
        if values[i] is not None:
            current.left = TreeNode(values[i])
            queue.append(current.left)
        i += 1
        if i < len(values) and values[i] is not None:
            current.right = TreeNode(values[i])
            queue.append(current.right)
        i += 1

    return root

# Example usage:
# Input: [2, 1, 3, None, None, 0, 1]
input_array = [2, 1, 3, None, None, 0, 1]
root = build_tree_from_list(input_array)
result = evaluateTree(root)
print(result)  # Output: True (for this example)
